File: iOS/SDK/smartsip-sdk/smartsip-sdk/Internal/linphone/CallState+linphone.swift
//
//  CallState+linphone.swift
//  smartsip-sdk
//
//  Created by Franz Iacob on 08/01/2026.
//

internal import linphonesw

extension CallState {
    /// Maps internal Linphone Call.State to the public SDK CallState
    static func from(linphoneState: Call.State) -> CallState? {
        switch linphoneState {
        case .OutgoingInit, .OutgoingProgress:
            return .dialing
            
        case .OutgoingRinging, .IncomingReceived, .PushIncomingReceived:
            return .ringing
            
        case .Connected, .StreamsRunning:
            return .connected
            
        case .Pausing, .Paused, .PausedByRemote:
            return .held
            
        case .End, .Released:
            return .disconnected
            
        case .Error:
            return .failed
            
        default:
            // Ignore internal technical transitions like 'Updating' or 'Referred'
            // unless you specifically want to handle them.
            return nil
        }
    }
    
    static func from(linphoneState: RegistrationState) -> CallState? {
        switch linphoneState {
        case .Ok:
            return .loggedIn
            
        case .Progress, .Refreshing:
            return .loginInProgress
    
        default:
            return .loggedOut
        }
    }
}

File: Android/smartsip-sdk/src/main/java/cc/smartconnect/smartsip_sdk/private/CallState+Linphone.kt
package cc.smartconnect.smartsip_sdk.private

import cc.smartconnect.smartsip_sdk.CallState
import org.linphone.core.Call
import org.linphone.core.RegistrationState

/**
 * CallState+Linphone.kt
 * smartsip-sdk
 * * Created by Franz Iacob on 21/01/2026.
 * * Extension factory methods to map Linphone-specific states to the SDK's
 * universal CallState representation.
 */

/**
 * Maps Linphone Call.State to SmartSip CallState.
 */
fun CallState.Companion.from(linphoneState: Call.State?): CallState? {
    return when (linphoneState) {
        Call.State.OutgoingInit,
        Call.State.OutgoingProgress -> CallState.DIALING

        Call.State.OutgoingRinging,
        Call.State.IncomingReceived -> CallState.RINGING

        Call.State.Connected,
        Call.State.StreamsRunning -> CallState.CONNECTED

        Call.State.Pausing,
        Call.State.Paused,
        Call.State.PausedByRemote -> CallState.HELD

        Call.State.End,
        Call.State.Released -> CallState.DISCONNECTED

        Call.State.Error -> CallState.FAILED

        else -> null
    }
}

/**
 * Maps Linphone RegistrationState to SmartSip CallState.
 */
fun CallState.Companion.from(registrationState: RegistrationState?): CallState? {
    return when (registrationState) {
        RegistrationState.Ok -> CallState.LOGGED_IN

        RegistrationState.Progress,
        RegistrationState.Refreshing -> CallState.LOGIN_IN_PROGRESS

        RegistrationState.Cleared,
        RegistrationState.None -> CallState.LOGGED_OUT

        RegistrationState.Failed -> CallState.FAILED

        else -> null
    }
}